pkg/repo/repo.go

```diff
@@ -127,36 +159,270 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+	}
+	var fulcioAuthority, tsaAuthority *v1.CertificateAuthority
+	var err error
+
+	// concat the fulcio chain and process it into CertificateAuthority
+	fulcioAuthority, err = certChainToAuthority(concatCertChain(fulcioLeaf, fulcioIntermed, fulcioRoot))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cert chain for Fulcio: %w", err)
+	}
+	tr.CertificateAuthorities = []*v1.CertificateAuthority{fulcioAuthority}
+
+	// concat the tsa chain and process it into CertificateAuthority
+	tsaAuthority, err = certChainToAuthority(concatCertChain(tsaLeaf, tsaIntermed, tsaRoot))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cert chain for TSA: %w", err)
+	}
+	tr.TimestampAuthorities = append(tr.TimestampAuthorities, tsaAuthority)
+
+	marshaller := &protojson.MarshalOptions{
+		Indent: "  ",
+	}
+	serialized, err := marshaller.Marshal(&tr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize trust root: %w", err)
+	}
+
+	return &TargetWithMetadata{
+		Name:  "trusted_root.json",
+		Bytes: serialized,
+	}, nil
+}
+
+func concatCertChain(leaf []byte, intermediate [][]byte, root []byte) []byte {
+	var result []byte
+	result = append(result, leaf...)
+	result = append(result, byte('\n'))
+	for _, intermed := range intermediate {
+		result = append(result, intermed...)
+		result = append(result, byte('\n'))
+	}
+	result = append(result, root...)
+	return result
+}
+
+func pubkeyToTLogInstance(key []byte) (*v1.TransparencyLogInstance, error) {
+	logId := sha256.Sum256(key)
+	der, _ := pem.Decode(key)
+	keyDetails, err := getKeyDetails(der.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.TransparencyLogInstance{
+		BaseUrl:       "",
+		HashAlgorithm: v1Common.HashAlgorithm_SHA2_256, // TODO: make it possible to change this value
+		PublicKey: &v1Common.PublicKey{
+			RawBytes:   der.Bytes,
+			KeyDetails: keyDetails,
+			ValidFor: &v1Common.TimeRange{
+				Start: timestamppb.New(time.Now()),
+			},
+		},
+		LogId: &v1Common.LogId{
+			KeyId: logId[:],
+		},
+	}, nil
+}
+
+func getKeyDetails(key []byte) (v1Common.PublicKeyDetails, error) {
+	var k any
+	var err1, err2 error
+
+	k, err1 = x509.ParsePKCS1PublicKey(key)
+	if err1 != nil {
+		k, err2 = x509.ParsePKIXPublicKey(key)
+		if err2 != nil {
+			return 0, fmt.Errorf("Can't parse public key with PKCS1 or PKIX: %w, %w", err1, err2)
+		}
+	}
+
+	// borrowed from https://github.com/kommendorkapten/trtool/blob/main/cmd/trtool/app/common.go
+	switch v := k.(type) {
+	case *ecdsa.PublicKey:
+		if v.Curve == elliptic.P256() {
+			return v1Common.PublicKeyDetails_PKIX_ECDSA_P256_SHA_256, nil
+		}
+		if v.Curve == elliptic.P384() {
+			return v1Common.PublicKeyDetails_PKIX_ECDSA_P384_SHA_384, nil
+		}
+		if v.Curve == elliptic.P521() {
+			return v1Common.PublicKeyDetails_PKIX_ECDSA_P521_SHA_512, nil
+		}
+		return 0, errors.New("unsupported elliptic curve")
+	case *rsa.PublicKey:
+		/*
+			NOTE: It is not possible to recognize padding from just the public key alone;
```

Comment:
I'd just omit this commented out block, almost all Sigstore clients can't handle RSA-PSS keys.